File: examples/python/archive_session.py
#!/usr/bin/env python3
"""Model Health Python SDK — session archive demo.

Walks through the archive workflow:
  1. Select a session
  2. Choose whether to include raw video files
  3. Request archive preparation
  4. Poll until the archive is ready (or fails)
  5. Download and save the ZIP file

Usage:
    archive_session.py [<api_key>]
"""

import sys
import time

from docopt import docopt

from modelhealth import (
    ModelHealthService,
    ModelHealthError,
    ArchiveStatus,
)
from _prompts import pick_one, confirm
from _utils import save_file, load_api_key

# ---------------------------------------------------------------------------
# Polling helper
# ---------------------------------------------------------------------------

def _poll_archive(service, archive, interval=2):
    """Block until the archive is ready or has failed.

    Returns the final status (ArchiveStatus.ready or ArchiveStatus.failed).
    """
    while True:
        status = service.archive_status(archive)
        if status == ArchiveStatus.processing:
            print("  Preparing archive...  ", end="\r", flush=True)
        else:
            print()  # clear the \r line
            return status
        time.sleep(interval)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    print("Connecting...")
    try:
        service = ModelHealthService(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")

    # Session
    print("\nFetching sessions...")
    sessions = service.session_list()
    if not sessions:
        sys.exit(
            "No sessions found. Create a session using the Model Health mobile app first."
        )

    print(f"\n{len(sessions)} session(s):\n")
    session = pick_one(
        sessions,
        "Select session to archive",
        lambda s: f"[session ID: {s.id}]  session name: {s.session_name or '(unnamed)'}  subject: {s.name or '(unnamed)'}  {s.activities_count} {'activity' if s.activities_count == 1 else 'activities'}",
    )

    # Options
    print()
    with_videos = confirm("Include raw video files in the archive?", default=False)

    # Request archive
    print(f"\nRequesting archive for session '{session.id}'...")
    try:
        archive = service.prepare_archive(session, with_videos=with_videos)
    except ModelHealthError as exc:
        sys.exit(f"Failed to start archive preparation: {exc}")

    # Poll for completion
    print("Waiting for archive to be ready...")
    status = _poll_archive(service, archive)

    if status != ArchiveStatus.ready:
        sys.exit(f"Archive preparation did not complete (status: {status}).")
    print("Archive is ready.")

    # Download and save
    print("\nDownloading...")
    try:
        data = service.archive_data(archive)
    except ModelHealthError as exc:
        sys.exit(f"Failed to download archive: {exc}")

    path = save_file(f"ModelHealth_Session_{session.id}.zip", data)
    print(f"  Saved {path}  ({len(data):,} bytes)")
    print("\nDone.")


if __name__ == "__main__":
    args = docopt(__doc__)
    main(load_api_key(args["<api_key>"]))

File: examples/swift/Sources/ArchiveSession/ArchiveSession.swift
/// Model Health Swift examples — session archive.
///
/// Usage:
///   swift run ArchiveSession [<api_key>]

import Foundation
import ModelHealth
import Shared

@main
struct ArchiveSession {
    static func main() async {
        let apiKey = loadAPIKey()
        print("Connecting...")
        let service: ModelHealthService
        do {
            service = try ModelHealthService(apiKey: apiKey)
        } catch {
            fputs("Failed to initialise: \(error)\n", stderr)
            exit(1)
        }

        print("\nFetching sessions...")
        let sessions: [Session]
        do {
            sessions = try await service.sessionList()
        } catch {
            fputs("Failed to fetch sessions: \(error)\n", stderr)
            exit(1)
        }

        guard !sessions.isEmpty else {
            fputs("No sessions found. Create a session using the Model Health mobile app first.\n", stderr)
            exit(1)
        }

        print("\n\(sessions.count) session(s):\n")
        let session = pickOne(
            from: sessions,
            prompt: "Select session to archive",
            label: { s in
                let sessionLabel = s.sessionName.isEmpty ? "(unnamed)" : s.sessionName
                let subjectLabel = s.name.isEmpty ? "(unnamed)" : s.name
                let actWord = s.activitiesCount == 1 ? "activity" : "activities"
                return "[session ID: \(s.id)]  session name: \(sessionLabel)  subject: \(subjectLabel)  \(s.activitiesCount) \(actWord)"
            }
        )

        print()
        let withVideos = confirm("Include raw video files in the archive?", default: false)

        print("\nRequesting archive for session '\(session.id)'...")
        let archive: Archive
        do {
            archive = try await service.prepareArchive(for: session, withVideos: withVideos)
        } catch {
            fputs("Failed to start archive preparation: \(error)\n", stderr)
            exit(1)
        }

        print("Waiting for archive to be ready...")
        let status = await pollArchive(service: service, archive: archive)

        guard case .ready = status else {
            fputs("Archive preparation did not complete (status: \(status)).\n", stderr)
            exit(1)
        }
        print("Archive is ready.")

        print("\nDownloading...")
        let data: Data
        do {
            data = try await service.archiveData(for: archive)
        } catch {
            fputs("Failed to download archive: \(error)\n", stderr)
            exit(1)
        }

        let path = saveFile(named: "ModelHealth_Session_\(session.id).zip", data: data)
        print("  Saved \(path)  (\(data.count.formatted()) bytes)")
        print("\nDone.")
    }
}

private func pollArchive(service: ModelHealthService, archive: Archive) async -> ArchiveStatus {
    while true {
        let status: ArchiveStatus
        do {
            status = try await service.archiveStatus(for: archive)
        } catch {
            fputs("Failed to check archive status: \(error)\n", stderr)
            exit(1)
        }

        if case .processing = status {
            print("  Preparing archive...  ", terminator: "\r")
            fflush(stdout)
        } else {
            print()
            return status
        }
        try? await Task.sleep(nanoseconds: 2_000_000_000)
    }
}
